src/subprocess-posix.cc

```diff
@@ -169,11 +170,61 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+bool SubprocessSet::child_exited_;
+
+void SubprocessSet::SetChildExited(int signum) {
+  child_exited_ = true;
+}
+
+void SubprocessSet::HandleChildExit() {
+  if (!child_exited_)
+    return;
+  child_exited_ = false;
+  for (int i = running_.size(); i > 0; --i)
```

Comment:
The idea behind that was that we know the maximum number of waitable children, so we could make that many waitpid calls without having it return with errno==ECHILD. I've changed it anyway, hopefully it's clearer now.